pytorch_lightning/trainer/supporters.py

```diff
@@ -119,27 +122,249 @@ def mean(self):
         return self.total / self.num_values
 
 
+class LightningBatchSamplerWrapper:
+    """
+    This class wraps user batch sampler, so we can extract
+    the batch_indices for tracking each sample.
+    """
+
+    def __init__(self, batch_sampler):
+        self.batch_sampler = batch_sampler
+        self.batch_indices = None
+
+    def __iter__(self):
+        for batch_indices in self.batch_sampler:
+            self.batch_indices = batch_indices
+            yield batch_indices
+
+    @staticmethod
+    def to_new_dataloader(dataloader: DataLoader) -> DataLoader:
```

Comment:
```suggestion
    def recreate_dataloader(dataloader: DataLoader) -> DataLoader:
```
I feel that the ord new is a bit confusing or we shall explain what the new means...